modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -189,8 +193,9 @@ AzureBlobServiceClient createClient(
             builder.credential(credentialBuilder.build());
         }
 
-        if (successfulRequestHandler != null) {
-            builder.addPolicy(new SuccessfulRequestTracker(purpose, successfulRequestHandler));
+        if (requestMetricsHandler != null) {
+            builder.addPolicy(new RequestMetricsTracker(purpose, requestMetricsHandler));
+            builder.addPolicy(RetryMetricsTracker.INSTANCE);
```

Comment:
`RequestMetricsTracker` sets up the context at the beginning of the request and dispatches the metrics to the listener at the end of the request, `RetryMetricsTracker` bumps the necessary counters for each retry (it's stateless and hence reusable).